collector/database/classes/collector.php

```diff
@@ -72,4 +88,40 @@ public function get_metrics($metricnames = null, $since = false): array {
                ORDER BY time asc";
         return $DB->get_records_sql($sql, $params);
     }
+
+    /**
+     * Records retrieved data in collector.
+     *
+     * @param  \tool_cloudmetrics\metric\base $metricclass Class representing metric.
+     * @param  array $data Array of metric items.
+     */
+    public function record_saved_metrics(\tool_cloudmetrics\metric\base $metricclass, array $metricitems = []) {
+        global $DB;
+
+        $maxbackdate = explode('-', $metricclass->get_range_retrieved())[0];
+        $transaction = $DB->start_delegated_transaction();
+        // Sets what data has been sent to collector.
+        $metricclass->set_data_sent_config();
+        if (is_array($metricitems) && count($metricitems) != 0 && !$metricclass->sameconfig) {
+            // When a further date is passed we only want to add missing data.
+            $iscompleted = ($metricitems[0]->time < $maxbackdate) ? true : false;
+            if ($iscompleted) {
+                // Trick to avoid erasing data when it only needs to be completed.
+                $this->record_metrics($metricitems);
+            } else {
+                $this->delete_metrics($metricclass->get_name());
```

Comment:
I'm a little confused why we'd ever need to delete data? If we are being asked to backfill data from time period which already has data then that smells like what ever is calling record_saved_metrics isn't sending correct times

Reply:
I found it confusing seeing different frequencies in the db - what happens at the moment : it retrieves all data for the requested time-frame unless the frequency changed, when frequency changes, it deletes all data and replaces it with the new data with proper interval. Also, if I request data for 4 months, it will fill data for those 4 months then if I request data for 12 months, it will retrieve data from the lowest timestamp present in DB to the passed point time requested (with the period) - it will only retrieve the missing part.

Reply:
I think deleting data has some business value but I think it should be explicit, similar to the setting for how much data you want to keep around. I don't like the idea of someone trying to add in more data and ending up losing data they already had.

I've split that off into a new issue, if you want you can tackle that after this pr lands:
https://github.com/catalyst/moodle-tool_cloudmetrics/issues/104

For this one I'd just want it to only ever add more data

Reply:
I have set it up so it never deletes data now - but also that if frequency changed it is not overlapping data of another interval - It would get extremely confusing. If frequency changes, timestamp need to be earlier than lowest one in DB so it completes data, it looks like this. Highlighted the one month period - the other one 1 week.

![Screenshot from 2022-06-08 15-54-27](https://user-images.githubusercontent.com/76186988/172705722-0e8d500b-8c04-43e7-9c03-a7217473ba4d.png)

I will sure get on this one once this PR lands.